src/r_mutex.rs

```diff
@@ -0,0 +1,122 @@
+//!
+
+use {
+    crate::{
+        error::{protect, Error},
+        object::Object,
+        ruby_sys::{
+            rb_mutex_lock, rb_mutex_locked_p, rb_mutex_new, rb_mutex_trylock, rb_mutex_unlock,
+        },
+        value::{private, NonZeroValue, ReprValue, Value},
+    },
+    std::{
+        fmt::{Debug, Display, Error as FError, Formatter},
+        ops::Deref,
+    },
+};
+
+/// A Value pointer to a RMutex struct, Ruby's internal representation of mutexes.
+///
+/// This is a lower level API where thread safety is not statically guaranteed.
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct RMutex(NonZeroValue);
+
+impl Deref for RMutex {
+    type Target = Value;
+    fn deref(&self) -> &Self::Target {
+        self.0.get_ref()
+    }
+}
+
+impl From<RMutex> for Value {
+    fn from(f: RMutex) -> Self {
+        *f
+    }
+}
+
+unsafe impl private::ReprValue for RMutex {
+    fn to_value(self) -> Value {
+        *self
+    }
+
+    unsafe fn from_value_unchecked(val: Value) -> Self {
+        Self(NonZeroValue::new_unchecked(val))
+    }
+}
+
+impl ReprValue for RMutex {}
+
+impl Object for RMutex {}
+
+impl Display for RMutex {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FError> {
+        write!(f, "{}", unsafe { self.to_s_infallible() })
+    }
+}
+
+impl Debug for RMutex {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FError> {
+        write!(f, "{}", self.inspect())
+    }
+}
+
+impl Default for RMutex {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl RMutex {
+    /// Create a new unlocked `RMutex`
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use magnus::{r_mutex::RMutex};
+    /// # let _cleanup = unsafe { magnus::embed::init() };
+    ///
+    /// let mutex = RMutex::new();
+    /// assert!(!mutex.locked());
+    ///
+    /// assert!(mutex.try_lock());
+    /// assert!(mutex.locked());
+    ///
+    /// mutex.unlock().unwrap();
+    /// assert!(!mutex.locked());
+    /// ```
+    pub fn new() -> Self {
```

Comment:
I think it would be better to have the sig be something like `RMutex<T>`, in which `.lock()` provides sync access to the inner `T`. Same way `std` does it. That will avoid having the imperative API. 

The benefits are: 
- no more forgetting to unlock since `Drop` can handle that
- can plug into `Sync` semantics in the type system
- don’t have to learn a new interface as a user

Reply:
I don’t think we should try and turn Ruby’s Mutex into a container like Rust’s Mutex, as that isn’t what the C/Ruby api is.

A guard value returned from lock, that unlocks on drop might be nice (kind of like how I did the embed::init and shutdown).

Reply:
> I don’t think we should try and turn Ruby’s Mutex into a container like Rust’s Mutex, as that isn’t what the C/Ruby api is.

Can you elaborate? For what it’s worth, I really like how magnus has diverged from the Ruby C API when it makes sense. In this case, it seems like magnus has a good opportunity to improve on imperative lock mechanisms which are historically brittle, usually too coarse, and hard to use correctly.

Reply:
Should we just wrap it using `BoxValue` + impl unsafe `Send` + `Sync`?

Im doing that on my project.